### PR TITLE
Compact field-join storage options, improved filterByField method

### PR DESCRIPTION
**Pull request by bergloman**
# Field-join binary representation

Field-joins can now be stored in more compact way. Develop can specify field-types to be used for storing the field-join's record id and frequency by providing `storage` tag:
~~~~~~~~~~~~~~~~~~~~~
joins: [{ name: 'parent', type: 'field', store: 'People', storage: 'int16-byte' }]
~~~~~~~~~~~~~~~~~~~~~
Default is `uint64-int`. It is also possible to set frequency type to empty string, which means that frequency will be always 1 and it wont take any space in the storage.

# filterByField
Recordset provides utility method `filterByField` that can now also operate on field joins:
~~~~~~~~~~~~~~~~~~~~~~~~~
recordset.filterByField("parent", parent_id, parent_id);
~~~~~~~~~~~~~~~~~~~~~~~~~
It currently accepts record ids. Caller can provide a range of IDs (min and max).

### CODE REVIEW

File: src/qminer/qminer_core.cpp [hunk 1]
```diff
@@ -2092,49 +2237,66 @@ PRecSet TRec::ToRecSet() const {
...
+                const int Pos = JoinIdPosH.GetKey(JoinId);
+                TMIn MIn(RecVal.GetBf() + Pos, RecVal.Len() - Pos, false);
+                JoinRecIdFqV.Load(MIn);
+            }
+        }
+        return TRecSet::New(JoinDesc.GetJoinStore(Base), JoinRecIdFqV);
+    } else if (JoinDesc.IsFieldJoin()) {
+        // do join using store field
+        const int JoinRecFieldId = JoinDesc.GetJoinRecFieldId();
+        const uint64 JoinRecId = IsFieldNull(JoinRecFieldId) ? TUInt64::Mx : GetFieldUInt64Safe(JoinRecFieldId);
```
Comment (blazf) [inline src/qminer/qminer_core.cpp]: Add function GetFieldJoinRecId to TStore that combines above two lines. Join provided as Id, Name, Desc.

File: src/qminer/qminer_core.cpp [hunk 2]
```diff
@@ -2146,39 +2308,43 @@ PRecSet TRec::DoJoin(const TWPt<TBase>& Base, const TStr& JoinNm) const {
...
 }
 
 TRec TRec::DoSingleJoin(const TWPt<TBase>& Base, const int& JoinId) const {
-	PRecSet JoinRecSet = DoJoin(Base, JoinId);
-	return TRec(JoinRecSet->GetStore(),
-		JoinRecSet->Empty() ? (uint64)TUInt64::Mx : JoinRecSet->GetRecId(0));
+    PRecSet JoinRecSet = DoJoin(Base, JoinId);
+    return TRec(JoinRecSet->GetStore(),
+        JoinRecSet->Empty() ? (uint64)TUInt64::Mx : JoinRecSet->GetRecId(0),
+        GetFieldJoinFq(JoinId));
```
Comment (blazf) [inline src/qminer/qminer_core.cpp]: Handle field joins directly, avoid PRecSet and DoJoin.

File: src/qminer/qminer_core.cpp [hunk 3]
```diff
@@ -2146,39 +2308,43 @@ PRecSet TRec::DoJoin(const TWPt<TBase>& Base, const TStr& JoinNm) const {
...
 }
 
 TRec TRec::DoSingleJoin(const TWPt<TBase>& Base, const TStr& JoinNm) const {
-	PRecSet JoinRecSet = DoJoin(Base, JoinNm);
-	return TRec(JoinRecSet->GetStore(),
-		JoinRecSet->Empty() ? (uint64)TUInt64::Mx : JoinRecSet->GetRecId(0));
+    PRecSet JoinRecSet = DoJoin(Base, JoinNm);
+    return TRec(JoinRecSet->GetStore(),
+        JoinRecSet->Empty() ? (uint64)TUInt64::Mx : JoinRecSet->GetRecId(0),
+        GetFieldJoinFq(Store->GetJoinId(JoinNm)));
```
Comment (blazf) [inline src/qminer/qminer_core.cpp]: Handle field joins directly, avoid PRecSet and DoJoin.

File: src/qminer/qminer_core.cpp [hunk 4]
```diff
@@ -744,6 +774,63 @@ void TStore::GetFieldTMem(const uint64& RecId, const int& FieldId, TMem& Mem) co
 PJsonVal TStore::GetFieldJsonVal(const uint64& RecId, const int& FieldId) const {
 	throw FieldError(FieldId, "Json");
 }
+*/
+
+/// Get field value using field id safely
+uint64 TStore::GetFieldUInt64Safe(const uint64& RecId, const int& FieldId) const {
+    switch (GetFieldDesc(FieldId).GetFieldType()) {
```
Comment (blazf) [inline src/qminer/qminer_core.cpp]: Make all casts to uint64 explicit.

File: src/qminer/qminer_core.cpp [hunk 5]
```diff
@@ -744,6 +774,63 @@ void TStore::GetFieldTMem(const uint64& RecId, const int& FieldId, TMem& Mem) co
...
+    case oftUInt16: return GetFieldUInt16(RecId, FieldId); break;
+    case oftUInt: return GetFieldUInt(RecId, FieldId); break;
+    case oftUInt64: return GetFieldUInt64(RecId, FieldId); break;
+    default: QmAssertR(false, TStr("GetFieldUInt64Safe: unsupported conversion for field id ") + FieldId);
+    }
+}
+
+/// Get field value using field id safely
+int64 TStore::GetFieldInt64Safe(const uint64& RecId, const int& FieldId) const {
+    switch (GetFieldDesc(FieldId).GetFieldType()) {
```
Comment (blazf) [inline src/qminer/qminer_core.cpp]: Make all casts to uint64 explicit.

File: src/qminer/qminer_core.cpp [hunk 6]
```diff
@@ -744,6 +774,63 @@ void TStore::GetFieldTMem(const uint64& RecId, const int& FieldId, TMem& Mem) co
...
+    case oftUInt16: return GetFieldUInt16(RecId, FieldId); break;
+    case oftUInt: return GetFieldUInt(RecId, FieldId); break;
+    case oftUInt64: return (int64)GetFieldUInt64(RecId, FieldId); break;
+    default: QmAssertR(false, TStr("GetFieldInt64Safe: unsupported conversion for field id ") + FieldId);
+    }
+}
+
+/// Set field value using field id (default implementation throws exception)
+void TStore::SetFieldUInt64Safe(const uint64& RecId, const int& FieldId, const uint64& UInt64) {
+    switch (GetFieldDesc(FieldId).GetFieldType()) {
```
Comment (blazf) [inline src/qminer/qminer_core.cpp]: Make all casts to uint64 explicit.

File: src/qminer/qminer_core.cpp [hunk 7]
```diff
@@ -744,6 +774,63 @@ void TStore::GetFieldTMem(const uint64& RecId, const int& FieldId, TMem& Mem) co
...
+    case oftUInt16: SetFieldUInt16(RecId, FieldId, (uint16)UInt64); break;
+    case oftUInt: SetFieldUInt(RecId, FieldId, (uint)UInt64); break;
+    case oftUInt64: SetFieldUInt64(RecId, FieldId, (uint64)UInt64); break;
+    default: QmAssertR(false, TStr("SetFieldUInt64Safe: unsupported conversion for field id ") + FieldId);
+    }
+}
+
+/// Set field value using field id (default implementation throws exception)
+void TStore::SetFieldInt64Safe(const uint64& RecId, const int& FieldId, const int64& Int64) {
+    switch (GetFieldDesc(FieldId).GetFieldType()) {
```
Comment (blazf) [inline src/qminer/qminer_core.cpp]: Make all casts to uint64 explicit.

File: src/qminer/qminer_core.cpp [hunk 8]
```diff
@@ -2092,49 +2237,66 @@ PRecSet TRec::ToRecSet() const {
 	return IsDef() ? TRecSet::New(Store, RecId) : TRecSet::New(Store);
 }
 
+int TRec::GetFieldJoinFq(const int& JoinId) const {
+    if (!Store->IsJoinId(JoinId)) {
+        return -1;
```
Comment (blazf) [inline src/qminer/qminer_core.cpp]: Change if to QmAssertR

File: src/qminer/qminer_core.cpp [hunk 9]
```diff
@@ -2092,49 +2237,66 @@ PRecSet TRec::ToRecSet() const {
...
+    }
+    const TJoinDesc& JoinDesc = Store->GetJoinDesc(JoinId);
+    return GetFieldJoinFq(JoinDesc);
+}
+
+int TRec::GetFieldJoinFq(const TJoinDesc& JoinDesc) const {
+    if (!JoinDesc.IsFieldJoin()) {
+        return -1;
```
Comment (blazf) [inline src/qminer/qminer_core.cpp]: Change if to QmAssertR

Comment (bergloman) [inline src/qminer/qminer_core.cpp]: Done

File: src/qminer/qminer_core.cpp (hunk 8 above)
Comment (bergloman) [inline src/qminer/qminer_core.cpp]: Done

File: src/qminer/qminer_core.cpp (hunk 7 above)
Comment (bergloman) [inline src/qminer/qminer_core.cpp]: Done

File: src/qminer/qminer_core.cpp (hunk 6 above)
Comment (bergloman) [inline src/qminer/qminer_core.cpp]: Done

File: src/qminer/qminer_core.cpp (hunk 5 above)
Comment (bergloman) [inline src/qminer/qminer_core.cpp]: Done

File: src/qminer/qminer_core.cpp (hunk 4 above)
Comment (bergloman) [inline src/qminer/qminer_core.cpp]: Done

File: src/qminer/qminer_core.cpp (hunk 3 above)
Comment (bergloman) [inline src/qminer/qminer_core.cpp]: Done

File: src/qminer/qminer_core.cpp (hunk 2 above)
Comment (bergloman) [inline src/qminer/qminer_core.cpp]: Done

File: src/qminer/qminer_core.cpp (hunk 1 above)
Comment (bergloman) [inline src/qminer/qminer_core.cpp]: Done

**Comment by bergloman** [comment]
Fixed all comments

**Pull request merged**